### ISSUE TITLE
Can we restrict to python 3.7 or later to take advantage of data classes?

### ISSUE THREAD

**Issue by DougBurke** [sherpa/sherpa#787]
We have a number of routines that return tuples that would benefit from some structure. Dataclasses, added in python 3.7, are a prime candidate for this (as better versions of NamedTuples), but it would require a minmum python of 3.7 (there is a compatibility module for python 3.6 so maybe it would just require dropping python 3.5 support). There are also other places where we could probably use dataclasses.

This would involve a number of breaking changes, so there's a usability hit above and beyond any development hit.

**Comment by hamogu** [comment]
Dataclasses are nice, but personally, I don't think they are such a great thing that I would break existing UI or API for it.

That said, python 3.9 will be out (https://www.python.org/dev/peps/pep-0602/) in Oct 2020. So, a Sherpa release, say in Dec 2020, could support 3.7 and 3.8 with experiemental support for 3.9. That's three major python version, which should be enough, so the Sherpa development version could switch to 3.7+ some time before that.

f-strings (introduced in python 3.6) are also great!

**Comment by hamogu** [comment]
By now, it's 2022 and we are in Python 3.7+, so we *could* use data classes. Personally, I prefer dicts if you just want some name- value mapping because dicts are more commonly used as input for other things and don't require a separate class definition. For example, it's easy to turn a list of dicts into a numpy structured array or and astropy Table, while a list of dataclass instances requires custom code.

Bottom line: I don't think we'll make is a priority to break or API and replace things with dataclasses just because we can, but if we do change things anyway, then dataclasses are available.

So, I'm closing this with "yes we can", but of course is can be re-opened if further discussion is warranted.
